chore: Simplify `PandasLikeDataFrame|DaskLazyFrame.join` method

FBruzzesi opened the pull request:
## What type of PR is this? (check all applicable)

- [x] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues

- Closes #2488

## Checklist

- [x] Code follows style guide (ruff)
- [ ] Tests added
- [ ] Documented the changes

## If you have comments or can explain your changes, please do so below

dangotbanned commented:
@FBruzzesi would you be open to flipping the method names around?

Most of the private ones that are related **share a common prefix**.
That helps them show up together in autocomplete 🙂 

I do understand why e.g (`anti_join`, `left_join`) make a lot of sense in isolation, but another way to think about it is:

```py
DataFrame.join(how="anti") -> CompliantDataFrame._join_anti
DataFrame.join(how="left") -> CompliantDataFrame._join_left
```

### `DaskLazyFrame`

![image](https://github.com/user-attachments/assets/60bafbce-a54c-4f82-a63c-b891e5c67fc4)

### `PandasLikeDataFrame`
![image](https://github.com/user-attachments/assets/a562a3cf-8f12-498a-b480-9b707df53531)

FBruzzesi commented:
@dangotbanned moved naming to `_join_<method>`

dangotbanned commented:
> @dangotbanned moved naming to `_join_<method>`

Thanks @FBruzzesi will try to do a proper review today 🙏

dangotbanned commented on narwhals/dataframe.py:
```diff
@@ -237,75 +237,66 @@ def sort(
-                right_on=right_on,
-                suffix=suffix,
+        if how == "cross":
+            if left_on is not None or right_on is not None or on is not None:
+                msg = "Can not pass `left_on`, `right_on` or `on` keys for cross join"
+                raise ValueError(msg)
+            result = compliant.join(
+                other, how=how, left_on=None, right_on=None, suffix=suffix
             )
-        )
+        elif on is None:
+            if left_on is None or right_on is None:
+                msg = f"Either (`left_on` and `right_on`) or `on` keys should be specified for {how}."
+                raise ValueError(msg)
+            if len(left_on) != len(right_on):
+                msg = "`left_on` and `right_on` must have the same length."
+                raise ValueError(msg)
+            result = compliant.join(
+                other, how=how, left_on=left_on, right_on=right_on, suffix=suffix
+            )
+        else:
+            if left_on is not None or right_on is not None:
+                msg = f"If `on` is specified, `left_on` and `right_on` should be None for {how}."
+                raise ValueError(msg)
+            result = compliant.join(
+                other, how=how, left_on=on, right_on=on, suffix=suffix
+            )
+        return self._with_compliant(result)
```
We've now got 3 distinct paths, which come from:

1. `how: Literal['cross'], on: None, left_on: None, right_on: None`
2. `how: Literal['inner', 'left', 'full', 'semi', 'anti'], on: None, left_on: list[str], right_on: list[str]`
3. `how: Literal['inner', 'left', 'full', 'semi', 'anti'], on: list[str], left_on: None, right_on: None`

Maybe we can utilize this to avoid the double-dipping like in (https://github.com/narwhals-dev/narwhals/pull/2000#discussion_r2083106551)?

dangotbanned replied:
## Edit
I need to remember my own advice 🤦‍♂️ 

me yesterday on discord 😄 
> Oh I see!
> So a few different ways to solve:
> 
> 1. Mutually exclusive `@overload`s
>   i. So you'd specify `None` is only allowed for both `*_on` when `how="cross"`
>   ii. However these can be tricky to get right, especially when there's a lot of parameters
> 2. Raise an error instead of asserting when an invariant is broken (https://github.com/narwhals-dev/narwhals/pull/2000/commits/529c196549c77f093c8a77a61e9ff26e9511eb66)
>   i. This is the **easiest solution**, but often *feels* like you're doing something unnecessary
> 3. Split incompatible signatures into distinct methods/functions
>   i. This is **my preferred solution** and what I'm usually hinting towards when I suggest adding multiple `@classmethod`s

dangotbanned replied:
A follow-up could look at option 3, which would be making some `CompliantJoin` class, then having:
- `CompliantJoin.left`
- `CompliantJoin.inner`
- ...

Would probably be easier to share code between `pandas` and `dask` that way as well 🙂

FBruzzesi replied:
> A follow-up could look at option 3, which would be making some `CompliantJoin` class, then having:
> 
> * `CompliantJoin.left`
> * `CompliantJoin.inner`
> * ...
> 
> Maybe the calls from `BaseFrame` would look like:
> 
> ```python
> compliant.join(**kwds).left()
> compliant.join(**kwds).inner()
> ```
> 
> Or just have the one `join`, which makes the other calls internally, to keep it simple from the outside:
> 
> ```python
> compliant.join(**kwds)
> self.left()
> ```
> 
> Would probably be easier to share code between `pandas` and `dask` that way as well 🙂

This seems like an interesting idea but I am not sure I get it right and/or I would end up with the expected changes.

Especially:

Where would `CompliantJoin` exist and how would it interact with other compliant classes?

and 

> Maybe the calls from `BaseFrame` would look like:
> 
> ```python
> compliant.join(**kwds).left()
> compliant.join(**kwds).inner()
> ```

Would this require some special path for polars?

dangotbanned replied:
@FBruzzesi you could think of it like these:
- `CompliantGroupBy`
- `CompliantWhen`
- #2393

Or similar to how the `Expr`, `Series` namespace classes work.

Essentially, we define some common parts and then refine from there per-backend

> Would this require some special path for polars?

Yeah most likely, `polars` gets to skip our abstractions in a lot of places and this would be one of those cases I'd assume.

Anyways, this is very much just the kernel of an idea 😅
The goal would be sharing more code between backends 😎

FBruzzesi replied:
Should we... keep that as a follow up then? 😅

> The goal would be sharing more code between backends 😎

Yep this is much loud and clear! Pandas-like and Dask can definitly share some functionalities eventually :)

dangotbanned replied:
> Should we... keep that as a follow up then? 😅

Yeah for sure!

I started with that in (https://github.com/narwhals-dev/narwhals/pull/2511#discussion_r2083257048) but clearly got carried away with my rambling 😄

dangotbanned commented on narwhals/_pandas_like/dataframe.py:
```diff
@@ -576,153 +576,185 @@ def group_by(
+        right_on_mapper = _remap_full_join_keys(left_on, right_on, suffix)
+        other_native = other.native.rename(columns=right_on_mapper)
+        check_column_names_are_unique(other_native.columns)
+        right_on = list(right_on_mapper.values())  # we now have the suffixed keys
+        return self._with_native(
+            self.native.merge(
+                other_native,
+                left_on=left_on,
+                right_on=right_on,
+                how="outer",
+                suffixes=("", suffix),
+            ),
+        )
```
```suggestion
        right_suffixed = list(right_on_mapper.values())
        return self._with_native(
            self.native.merge(
                other_native,
                left_on=left_on,
                right_on=right_suffixed,
                how="outer",
                suffixes=("", suffix),
            )
        )
```

dangotbanned commented on narwhals/_pandas_like/dataframe.py:
```diff
@@ -576,153 +576,185 @@ def group_by(
+            key_token = generate_temporary_column_name(
+                n_bytes=8, columns=[*self.columns, *other.columns]
             )
+
```
```suggestion
```

dangotbanned commented on narwhals/_pandas_like/dataframe.py:
```diff
@@ -576,153 +576,185 @@ def group_by(
                 )
+                .drop(columns=key_token)
             )
 
```
```suggestion
```

dangotbanned commented on narwhals/_pandas_like/dataframe.py:
```diff
@@ -576,153 +576,185 @@ def group_by(
+    def _join_anti(
+        self, other: Self, *, left_on: Sequence[str], right_on: Sequence[str]
+    ) -> Self:
+        implementation = self._implementation
+        backend_version = self._backend_version
+
+        if implementation.is_cudf():
             return self._with_native(
                 self.native.merge(
-                    other_native,
-                    left_on=left_on,
-                    right_on=right_on,
-                    how="outer",
-                    suffixes=("", suffix),
-                ),
+                    other.native, how="leftanti", left_on=left_on, right_on=right_on
+                )
             )
 
+        indicator_token = generate_temporary_column_name(
+            n_bytes=8, columns=[*self.columns, *other.columns]
+        )
+
+        # rename to avoid creating extra columns in join
+        other_native = rename(
+            select_columns_by_name(
+                other.native,
+                column_names=list(right_on),
+                backend_version=backend_version,
+                implementation=implementation,
+            ),
+            columns=dict(zip(right_on, left_on)),
+            implementation=implementation,
+            backend_version=backend_version,
+        ).drop_duplicates()
+
         return self._with_native(
             self.native.merge(
-                other.native,
+                other_native,
```
```suggestion
    def _join_filter_rename(
        self, other: Self, left_on: Sequence[str], right_on: Sequence[str]
    ) -> pd.DataFrame:
        """Rename to avoid creating extra columns in `"anti"`, `"semi"` join, and avoids potential rows duplication."""
        implementation = self._implementation
        backend_version = self._backend_version
        return rename(
            select_columns_by_name(
                other.native, list(right_on), backend_version, implementation
            ),
            columns=dict(zip(right_on, left_on)),
            implementation=implementation,
            backend_version=backend_version,
        ).drop_duplicates()

    def _join_semi(
        self, other: Self, *, left_on: Sequence[str], right_on: Sequence[str]
    ) -> Self:
        other_native = self._join_filter_rename(other, left_on, right_on)
        return self._with_native(
            self.native.merge(
                other_native, how="inner", left_on=left_on, right_on=left_on
            )
        )

    def _join_anti(
        self, other: Self, *, left_on: Sequence[str], right_on: Sequence[str]
    ) -> Self:
        if self._implementation.is_cudf():
            return self._with_native(
                self.native.merge(
                    other.native, how="leftanti", left_on=left_on, right_on=right_on
                )
            )
        indicator_token = generate_temporary_column_name(
            n_bytes=8, columns=[*self.columns, *other.columns]
        )
        return self._with_native(
            self.native.merge(
                self._join_filter_rename(other, left_on, right_on),
```

dangotbanned replied:
What on earth is a `join_filter` I hear you ask? 😉 

The last two are not like the others, which is why they require one less parameter

https://github.com/narwhals-dev/narwhals/blob/dd41833963acaf2641c6c4c8948ee7aa5f25734a/narwhals/typing.py#L239-L250

FBruzzesi replied:
Solved in https://github.com/narwhals-dev/narwhals/pull/2511/commits/a0fca3d5c7bcfaa67c870b7e1c384bafa477ec04

dangotbanned commented on narwhals/_pandas_like/dataframe.py:
```diff
@@ -576,153 +576,185 @@ def group_by(
+            right_on=right_on,
+            suffixes=("", suffix),
+        )
+        extra = []
+        for left_key, right_key in zip(left_on, right_on):
+            if right_key != left_key and right_key not in self.columns:
+                extra.append(right_key)
+            elif right_key != left_key:
+                extra.append(f"{right_key}{suffix}")
```
Might there be a way to do this using `set` operations?

We don't need the order in the `drop(columns=...)` call

FBruzzesi replied:
Solved in https://github.com/narwhals-dev/narwhals/pull/2511/commits/ae64d74df150a27db63a6a3fc9a1e9ceecc832ac

dangotbanned reviewed (commented):
Thanks @FBruzzesi - looking a lot better! 🥳 

I've got a few suggestions on the `pandas` one, but I think the same would apply for `dask` as well

FBruzzesi commented:
> Thanks @FBruzzesi - looking a lot better! 🥳
> 
> I've got a few suggestions on the `pandas` one, but I think the same would apply for `dask` as well

Thanks @dangotbanned - great work you've done!
I will have time later today to go through each point

dangotbanned commented:
I was thinking about simplifying `BaseFrame.join_asof` in a similar way to (https://github.com/narwhals-dev/narwhals/pull/2511/commits/1c1daae80ba81597ce117fa8e055d0053354991a)

I thought I'd take a look at the `polars` version(S).

They're simpler, *and* accept more parameters and types 🤔 

- https://github.com/pola-rs/polars/blob/34b5729cec09a8e3619c70844a77830c0613f070/py-polars/polars/dataframe/frame.py#L7614-L7630
- https://github.com/pola-rs/polars/blob/34b5729cec09a8e3619c70844a77830c0613f070/py-polars/polars/lazyframe/frame.py#L5283-L5319

Maybe they simplfied it at some point?

FBruzzesi commented on narwhals/_dask/dataframe.py:
```diff
@@ -258,133 +258,160 @@ def sort(
+        )
+        df = self.native.merge(
+            other_native,
+            how="outer",
```
Spotted that we are wasted some cycles - this can be a left join (with the same filter below). Fixing that

FBruzzesi commented on narwhals/_pandas_like/dataframe.py:
```diff
@@ -563,153 +563,185 @@ def group_by(
-                left_on=left_on,
-                right_on=right_on,
-                suffixes=("", suffix),
+    def _join_semi(
+        self, other: Self, *, left_on: Sequence[str], right_on: Sequence[str]
+    ) -> Self:
+        other_native = self._join_filter_rename(
+            other=other,
+            columns_to_select=list(right_on),
+            columns_mapping=dict(zip(right_on, left_on)),
+        )
```
@dangotbanned please forgive me - I didn't get your point here 🥲

dangotbanned commented on narwhals/_pandas_like/dataframe.py:
```diff
@@ -563,153 +563,185 @@ def group_by(
+            .drop(columns=indicator_token)
         )
 
+    def _join_filter_rename(
+        self, other: Self, columns_to_select: list[str], columns_mapping: dict[str, str]
+    ) -> pd.DataFrame:
+        """Helper function to avoid creating extra columns and row duplication.
```
@FBruzzesi in (https://github.com/narwhals-dev/narwhals/pull/2511#discussion_r2105813874) I'm saying that this version is more duplicated than what I had in the original suggestion, which was: 

```py
    def _join_filter_rename(
        self, other: Self, left_on: Sequence[str], right_on: Sequence[str]
    ) -> pd.DataFrame:
        """Rename to avoid creating extra columns in `"anti"`, `"semi"` join, and avoids potential rows duplication."""
        implementation = self._implementation
        backend_version = self._backend_version
        return rename(
            select_columns_by_name(
                other.native, list(right_on), backend_version, implementation
            ),
            columns=dict(zip(right_on, left_on)),
            implementation=implementation,
            backend_version=backend_version,
        ).drop_duplicates()
```

The `list(right_on)` and `dict(zip(right_on, left_on))` parts are defined in one place - rather than repeated each time you call `_join_filter_rename`

FBruzzesi replied:
Oh I see, it has been a while since this PR, I didn't spot it in the other comment.
I see what you mean, it a similar conundrum to #2495, the signature with `columns_to_select` and `columns_mapping` in my opinion is much more intuitive to understand what happens within the method without jumping into it, on the other hand, yes, we end up creating and passing these objects externally of the method itself.
I really don't know 🤔

dangotbanned commented:
https://github.com/narwhals-dev/narwhals/pull/2511#event-18343540555

@FBruzzesi you may have noticed I remembered about this PR in (https://github.com/narwhals-dev/narwhals/pull/2511#ref-pullrequest-3165108714) and then quickly lost track 😅 

Feel free to ping me if I forget to review tomorrow!

dangotbanned reviewed (changes requested):
Thanks for the PR @FBruzzesi (*and bearing with me* 😄)

I've got a couple of new notes, which are what the *Request changes* are related to.

Regarding (https://github.com/narwhals-dev/narwhals/pull/2511#discussion_r2105922818), I still have a preference for my earlier suggestion.
*But*, **both** are improvements over what we have on `main` - so feel free to keep that part as-is 🙂 

![](https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExaXF1NG1nZWk1cnhvZHoxbzBmY2x5ZWFvcHZzZ2xqdzV3d3FhN211cyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/eEA8wW82Ae7Ep99dCg/giphy.gif)

FBruzzesi commented on narwhals/_pandas_like/dataframe.py:
```diff
@@ -557,153 +557,185 @@ def group_by(
+            .drop(columns=indicator_token)
         )
 
+    def _join_filter_rename(
+        self, other: Self, columns_to_select: list[str], columns_mapping: dict[str, str]
+    ) -> pd.DataFrame:
+        """Helper function to avoid creating extra columns and row duplication.
+
+        Used in `"anti"` and `"semi`" join's.
+
+        Notice that a native object is returned.
+        """
```
Ok just kidding, getting headaches with the type checker (rightfully so) because of:

```diff
if how == "cross":
-     return self._join_cross(other=other, suffix=suffix)
+     result = self._join_cross(other=other, suffix=suffix)

if left_on is None or right_on is None:  # pragma: no cover
    raise ValueError(left_on, right_on)
```
now the second condition is always checked

dangotbanned replied:
@FBruzzesi sorry!!!!! I edited your comment instead of replying by mistake 😬 

How about this (like in `concat` 😉 )
```diff
if how == "cross":
-     return self._join_cross(other=other, suffix=suffix)
+     result = self._join_cross(other=other, suffix=suffix)
- if left_on is None or right_on is None:  # pragma: no cover
+ elif left_on is None or right_on is None:  # pragma: no cover
    raise ValueError(left_on, right_on)
```

dangotbanned commented on narwhals/_pandas_like/dataframe.py:
```diff
@@ -557,153 +557,185 @@ def group_by(
             self.native.merge(
-                other.native,
+                other_native,
+                # TODO(FBruzzesi): Raise issue upstream for Modin
```
Thanks!

So is it that `how="outer"` isn't supported for `modin`?

dangotbanned replied:
Ah https://github.com/modin-project/modin/issues/7384

Nice one thank you ❤️

FBruzzesi replied:
Found and linked a very similar issue: https://github.com/modin-project/modin/issues/7384

FBruzzesi commented:
Thanks for all the support @dangotbanned 
I (should) have addressed all your points. Only discussion left is:

> Regarding (https://github.com/narwhals-dev/narwhals/pull/2511#discussion_r2105922818), I still have a preference for my earlier suggestion.

but as of:

> But, both are improvements over what we have on main - so feel free to keep that part as-is 🙂

for now yes I would rather keep it as is. My mind has not changed from the [previous comment on it](https://github.com/narwhals-dev/narwhals/pull/2511#discussion_r2105926752) 😂

dangotbanned reviewed (approved):
![](https://64.media.tumblr.com/64fb369db15c1b058a5335f8f9ae24ba/tumblr_oql8iwJISo1qak30bo1_500.gifv)
